feat($goQuery): Experimental integration of Query

mattcreager opened the pull request:
$goQuery is a service built on `key.query()`.  It accepts a key, room,
filter and options.  It returns a QueryModel. The Model has the standard
`$sync` method which keeps the models value current with that of the
query result set.

``` js
angular.module('ChatApp').controller('ChatCtrl', ['$scope', '$goQuery', function($scope, $query) {
  var filter = { sender: { '$eq': 'GrandMasterLivingston' }};
  var sort = { $name: 'desc' };
  var limit = 20;
  var options = { sort: sort, limit: limit };

  $scope.chat = $query('messages', filter, options);
  $scope.chat.$sync();
}]);
```

Related: #48

Commit:
feat($goQuery): Experimental intergration of Query

$goQuery is a service built on `key.query()`.  It accepts a key, room,
filter and options.  It returns a QueryModel. The Model has the standard
`$sync` method which keeps the models value current with that of the
query result set.

```js
	angular.module('ChatApp').controller('ChatCtrl', ['$scope', '$goQuery', function($scope, $query) {
	  var filter = { “sender”: { “$eq”: “GrandMasterLivingston” }};
	  var sort = { “$name”: “desc”};
	  var limit = 20;
	  var options = { sort: sort, limit: limit };

	  $scope.chat = $query('messages', filter, options);
	  $scope.chat.$sync();
	}]);
```

Related: #48

ianlivingstone commented on lib/util/args.js:
```diff
@@ -0,0 +1,306 @@
+/**
```
cant you use bower to install this? why does it need to be a checked-in dependency

mattcreager replied:
I'd love to check this in with bower, but we're using component, which this library doesn't offer... I'd fork & and add it, but we're planning to over-haul the build process anyway!

We could also just manage the argument validation, but I like how this library is documented, implemented, and I think it's one positive step towards a cleaner interface.

Commit:
docs(query.md): Adding query documentation draft

Commit:
docs(key.md): Updating for consistencies sake

Commit:
chore(model): Refactored to enable sync replacement

Commit:
test(sync, model): Moving tests towards est. interface

Cleaning up tests, adding missing omit test, stripping
out anything DI related, we could consider dropping this
into an intergration test at some point.

Commit:
tests(query): Adding query sync tests

Commit:
tests(query): Adding query sync tests

colinmacdonald commented on docs/v1/query.md:
```diff
@@ -0,0 +1,59 @@
+| keyName |
+|:---|
+| Type: [String](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String). |
+| The name of the key to be referenced. [Naming restrictions apply.](Naming restrictions apply) A leading forward-slash ('/') is implied, (e.g. 'foo/bar' is the same as '/foo/bar'). |
```
Replace "Naming restrictions apply" with relative link to docs.

colinmacdonald commented on tests/query_sync.js:
```diff
@@ -0,0 +1,208 @@
+          context: { position: { current: 0 } },
+          expect: { foo: 'bar' }
+        },
+        // {
```
don't forget to uncomment these tests.

colinmacdonald commented on lib/query_sync.js:
```diff
@@ -0,0 +1,129 @@
+
+  /**
+   * @public
+   * @param {Object} key - GoInstant key
```
Inline docs - should be a "query" object, not a key?

Commit:
Merge remote-tracking branch 'origin/develop' into mattcreager/query

Conflicts:
	examples/index.html
	examples/js/example.js

colinmacdonald commented:
LGTM when :cactus:

Commit:
Merge branch 'mattcreager/query' of github.com:goinstant/goangular into mattcreager/query

Commit:
docs(query): Fixing IE8 Issue :|
